cmd/geth/dbcmd.go

```diff
@@ -929,3 +982,345 @@ func inspectHistory(ctx *cli.Context) error {
 	}
 	return inspectStorage(triedb, start, end, address, slot, ctx.Bool("raw"))
 }
+
+// pruneHistory implements the 'db prune-history' command, truncating the freezer
+// at the merge block.
+func pruneHistory(ctx *cli.Context) error {
+	stack, _ := makeConfigNode(ctx)
+	defer stack.Close()
+
+	db := utils.MakeChainDatabase(ctx, stack, false)
+	defer db.Close()
+
+	// Find the merge block (first block with zero difficulty)
+	log.Info("Looking for the merge block...")
+
+	// Get the current head block
+	headHash := rawdb.ReadHeadBlockHash(db)
+	if headHash == (common.Hash{}) {
+		return fmt.Errorf("chain head not found")
+	}
+
+	headNumber := rawdb.ReadHeaderNumber(db, headHash)
+	if headNumber == nil {
+		return fmt.Errorf("header number of head block not found")
+	}
+
+	// Binary search to find the merge block
+	var (
+		low        = uint64(0)
+		high       = *headNumber
+		mergeBlock = uint64(0)
+		found      = false
+	)
+
+	log.Info("Searching for merge block using binary search", "head", *headNumber)
+
+	for low <= high {
+		mid := (low + high) / 2
+		header := rawdb.ReadHeader(db, rawdb.ReadCanonicalHash(db, mid), mid)
+		if header == nil {
+			return fmt.Errorf("header %d not found", mid)
+		}
+
+		if header.Difficulty.Sign() == 0 {
+			// This is a post-merge block, look earlier
+			high = mid - 1
+			mergeBlock = mid
+			found = true
+		} else {
+			// This is a pre-merge block, look later
+			low = mid + 1
+		}
+	}
```

Comment:
You could do something like this here
```suggestion
    sort.Search(*headNumber, func(index int) bool {
         header := rawdb.ReadHeader(db, rawdb.ReadCanonicalHash(db, index), index)
         if header == nil {
             panic(fmt.Sprintf("header %d not found", index))
         }
         return header.Difficulty.Sign() == 0
         })
```

Reply:
Can't we just assume the merge block is hardcoded, and assume that we are running on a pre-merged chain?